### PR TITLE
Print environment variables for `cargo run` in extra verbose mode

### PR DESCRIPTION
**Pull request by Kobzol**
### What does this PR try to resolve?

This PR changes `cargo run` so that it prints the environment variables that it sets for the executed binary in extra verbose mode (`cargo run -vv`). This is particularly useful when it sets `LD_LIBRARY_PATH` which is needed to actually run the binary. Because if users want to run the binary without `cargo`, they will need to know where should they point that variable. This is described in issue https://github.com/rust-lang/cargo/issues/4879.

Fixes: https://github.com/rust-lang/cargo/issues/4879

### How should we test and review this PR?

I tried to write a functional test, but I'm not sure how to check that `stderr` contains only a given substring, not a whole line.

### CODE REVIEW

File: tests/testsuite/run.rs [hunk 1]
```diff
@@ -1416,6 +1416,23 @@ fn default_run_workspace() {
...
+            r#"
+                [package]
+                name = "a"
+                version = "0.0.1"
+            "#,
+        )
+        .file("src/main.rs", r#"fn main() {println!("run-a");}"#)
+        .build();
+
+    p.cargo("run -vv").with_stderr("LD_LIBRARY_PATH").run();
```
Comment (weihanglo) [inline tests/testsuite/run.rs:1433]: This contains two shortcomings:

* [Different platforms have different library paths](https://doc.rust-lang.org/nightly/cargo/reference/environment-variables.html#dynamic-library-paths). This won't pass on Windows and macOS.
* This matches both `rustc` and the final binary invocations. We can't tell whether the env is from which.

What about this? `[..]` for wildcard matches, and we match one env vars to verify the behavior.

```suggestion
    p.cargo("run -vv")
        .with_stderr(
            "\
[COMPILING] a v0.0.1 ([CWD])
[RUNNING] `[..]CARGO_MANIFEST_DIR=[CWD][..] rustc --crate-name a[..]`
[FINISHED] dev [unoptimized + debuginfo] target(s) in [..]
[RUNNING] `[..]CARGO_MANIFEST_DIR=[CWD][..] target/debug/a[EXE]`",
        )
        .run();
```

**Review by weihanglo — COMMENTED** [review]
Looks good to me. Do you want to deal with `cargo test` as well in this pull request, or a follow-up?

(Can also leave it there. I am fine with that 😁)

File: tests/testsuite/run.rs (hunk 1 above)
Comment (arlosi) [inline tests/testsuite/run.rs:1433]: `LD_LIBRARY_PATH` won't be set on Windows. I'd recommend a different one like `CARGO_PKG_NAME` for the test case. 

You can use `[..]` as a wildcard for matching.

```suggestion
    p.cargo("run -vv")
        .with_stderr_contains("[RUNNING] [..]CARGO_PKG_NAME=a[..]")
        .run();
```

**Review by arlosi — CHANGES_REQUESTED** [review]

File: tests/testsuite/run.rs (hunk 1 above)
Comment (arlosi) [reply tests/testsuite/run.rs:1433]: @weihanglo beat me to the review with the same comment 😊. I'll close this one.

**Comment by Kobzol** [comment]
Thanks for the hints. I added support for `cargo test` too. I have to say, it would be really nice if `cargo` had auto-blessable tests with placeholders :laughing: Getting the text output right is not straightforward.

**Comment by weihanglo** [comment]
I believe changing to `[RUNNING] `[..] rustdoc [..]--test [..]` would fix CI failures.

**Comment by Kobzol** [comment]
Thanks for the hint! I fixed the two failing tests and used `[DOCTEST] foo` in the `test` test, as I noticed it was used in the failing tests. I also added a test for `bench`.

**Comment by arlosi** [comment]
Thanks!

@bors r+

**Review by arlosi — APPROVED** [review]

**Pull request merged**